### PR TITLE
gh-117841: Add C implementation of `ntpath.lexists`

### PR DESCRIPTION
**Pull request by nineteendo**
### Benchmark

<details><summary>script</summary>

```bat
::speedup-ntpath.lexists.bat
@echo off
echo existent && call main\python -m timeit -s "import ntpath" "ntpath.lexists('main')" && call speedup-ntpath.lexists\python -m timeit -s "import ntpath" "ntpath.lexists('main')"
echo non-existent && call main\python -m timeit -s "import ntpath" "ntpath.lexists('foo')" && call speedup-ntpath.lexists\python -m timeit -s "import ntpath" "ntpath.lexists('foo')"
```
</details>

```none
existent
5000 loops, best of 5: 46.5 usec per loop # before
10000 loops, best of 5: 33.2 usec per loop # after
# -> 1.40x faster
non-existent
10000 loops, best of 5: 22.2 usec per loop # before
20000 loops, best of 5: 15.6 usec per loop # after
# -> 1.42x faster
```

* Issue: gh-117841

### CODE REVIEW

**Comment by nineteendo** [comment]
@serhiy-storchaka, is this better? This speeds us `ntpath.lexists()` without exposing a new parameter for `ntpath.exists()`.

**Comment by nineteendo** [comment]
@barneygale, do you think this can be merged? Or do you want more benchmarks?

**Comment by eryksun** [comment]
I'm working on what I hope will be an improved version compared to the first draft. AFAIK, the `GetFileInformationByName()` fast path will be generally available when Windows 11 24H2 is released later this year. Until then, and for older systems, we need to focus on improving the 'slow' path. I'm looking to avoid full `STAT` and `LSTAT` calls, accept as a last resort.

**Comment by eryksun** [comment]
Here's the revised implementation of `nt_exists()`:

```c
static PyObject *
nt_exists(PyObject *path, int follow_symlinks)
{
    path_t _path = PATH_T_INITIALIZE("exists", "path", 0, 1);
    HANDLE hfile;
    int result = 0;

    if (!path_converter(path, &_path)) {
        path_cleanup(&_path);
        if (PyErr_ExceptionMatches(PyExc_ValueError)) {
            PyErr_Clear();
            Py_RETURN_FALSE;
        }
        return NULL;
    }

    Py_BEGIN_ALLOW_THREADS
    if (_path.fd != -1) {
        hfile = _Py_get_osfhandle_noraise(_path.fd);
        if (hfile != INVALID_HANDLE_VALUE) {
            result = 1;
        }
    }
    else if (_path.wide) {
        BOOL slow_path = TRUE;
        FILE_STAT_BASIC_INFORMATION statInfo;
        if (_Py_GetFileInformationByName(_path.wide, FileStatBasicByNameInfo,
                &statInfo, sizeof(statInfo)))
        {
            if (!(statInfo.FileAttributes & FILE_ATTRIBUTE_REPARSE_POINT) ||
                !follow_symlinks &&
                IsReparseTagNameSurrogate(statInfo.ReparseTag))
            {
                slow_path = FALSE;
                result = 1;
            }
        }
        else if (_Py_GetFileInformationByName_ErrorIsTrustworthy(
                        GetLastError()))
        {
            slow_path = FALSE;
        }
        if (slow_path) {
            BOOL traverse = follow_symlinks;
            if (!traverse) {
                hfile = CreateFileW(_path.wide, FILE_READ_ATTRIBUTES, 0, NULL,
                            OPEN_EXISTING, FILE_FLAG_OPEN_REPARSE_POINT |
                            FILE_FLAG_BACKUP_SEMANTICS, NULL);
                if (hfile != INVALID_HANDLE_VALUE) {
                    FILE_ATTRIBUTE_TAG_INFO info;
                    if (GetFileInformationByHandleEx(hfile,
                            FileAttributeTagInfo, &info, sizeof(info)))
                    {
                        if (!(info.FileAttributes &
                                FILE_ATTRIBUTE_REPARSE_POINT) ||
                            IsReparseTagNameSurrogate(info.ReparseTag))
                        {
                            result = 1;
                        }
                        else {
                            // reparse point but not name-surrogate
                            traverse = TRUE;
                        }
                    }
                    else {
                        // legacy filesystem
                        traverse = TRUE;
                    }
                    CloseHandle(hfile);
                }
                else {
                    STRUCT_STAT st; 
                    switch (GetLastError()) {
                    case ERROR_ACCESS_DENIED:
                    case ERROR_SHARING_VIOLATION:
                    case ERROR_CANT_ACCESS_FILE:
                    case ERROR_INVALID_PARAMETER:
                        if (!LSTAT(_path.wide, &st)) {
                            result = 1;
                        }
                    }
                }
            }
            if (traverse) {
                hfile = CreateFileW(_path.wide, FILE_READ_ATTRIBUTES, 0, NULL,
                            OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
                if (hfile != INVALID_HANDLE_VALUE) {
                    CloseHandle(hfile);
                    result = 1;
                }
                else {
                    STRUCT_STAT st; 
                    switch (GetLastError()) {
                    case ERROR_ACCESS_DENIED:
                    case ERROR_SHARING_VIOLATION:
                    case ERROR_CANT_ACCESS_FILE:
                    case ERROR_INVALID_PARAMETER:
                        if (!STAT(_path.wide, &st)) {
                            result = 1;
                        }
                    }
                }
            }
        }
    }
    Py_END_ALLOW_THREADS

    path_cleanup(&_path);
    if (result) {
        Py_RETURN_TRUE;
    }
    Py_RETURN_FALSE;
}
```

**Comment by nineteendo** [comment]
Are you happy with the new implementation, or do you have more ideas?

**Comment by nineteendo** [comment]
The performance for existent files has already greatly improved. Nice job!

File: Modules/posixmodule.c [hunk 1]
```diff
@@ -5366,6 +5400,40 @@ os__path_exists_impl(PyObject *module, PyObject *path)
 }
 
 
+/*[clinic input]
+os._path_exists
+
+    path: 'O'
```
Comment (serhiy-storchaka) [inline Modules/posixmodule.c]:
```suggestion
    path: object
    /
```

File: Modules/posixmodule.c [hunk 2]
```diff
@@ -5366,6 +5400,40 @@ os__path_exists_impl(PyObject *module, PyObject *path)
...
+
+Test whether a path exists.  Returns False for broken symbolic links
```
Comment (serhiy-storchaka) [inline Modules/posixmodule.c]:
```suggestion
Test whether a path exists.  Returns False for broken symbolic links.
```

File: Modules/posixmodule.c [hunk 3]
```diff
@@ -5366,6 +5400,40 @@ os__path_exists_impl(PyObject *module, PyObject *path)
...
+/*[clinic end generated code: output=f508c3b35e13a249 input=380f77cdfa0f7ae8]*/
+{
+    return nt_exists(path, 1);
+}
+
+
+/*[clinic input]
+os._path_lexists
+
+    path: 'O'
```
Comment (serhiy-storchaka) [inline Modules/posixmodule.c]:
```suggestion
    path: object
    /
```

File: Modules/posixmodule.c [hunk 4]
```diff
@@ -5366,6 +5400,40 @@ os__path_exists_impl(PyObject *module, PyObject *path)
...
+
+Test whether a path exists.  Returns True for broken symbolic links
```
Comment (serhiy-storchaka) [inline Modules/posixmodule.c]:
```suggestion
Test whether a path exists.  Returns True for broken symbolic links.
```

File: Lib/test/test_genericpath.py [hunk 5]
```diff
@@ -128,31 +128,34 @@ def test_filetime(self):
         )
 
     def test_exists(self):
+        def check_exists(filename, expected):
```
Comment (serhiy-storchaka) [inline Lib/test/test_genericpath.py]: I do not think that this makes the code simpler. It also makes difficult to see what was added or changed in these tests. Should not it be pure optimization?

**Review by serhiy-storchaka — COMMENTED** [review]
At first glance looks correct, although I am not happy that we added so much complicated code for pure optimization of functions which should not be used in performance critical code (`os.scandir()` should be used instead).

**Comment by eryksun** [comment]
> At first glance looks correct, although I am not happy that we added so much complicated code for pure optimization of functions which should not be used in performance critical code (`os.scandir()` should be used instead).

The performance of `os.stat()` and `os.lstat()` will improve on Windows once `GetFileInformationByName()` is generally supported, starting with Windows 11 24H2 later this year. Eventually I expect the builtin `_path_*` tests to be removed when Python stops supporting Windows 10.

A disappointment with these builtin functions is that they can't be leveraged in `pathlib.Path` because they're designed to simply return `True` or `False` without raising `OSError` exceptions. The problem is that the `pathlib.Path` methods `exists()`, `is_dir()`, `is_file()`, and `is_symlink()` only ignore `OSError` exceptions for a small set of `errno` values and Windows error codes.

Regarding `scandir()`, in principle, the C implementations of `_path_isdir()` and `_path_isfile()` could be used by the `DirEntry` methods `is_dir()` and `is_file()`. These methods have to call `STAT()` if the entry is a reparse point, except if `follow_symlinks` is false and it's a name-surrogate reparse point[^1]. That's more work than necessary. However, the `DirEntry` tests only handle `FileNotFoundError` exceptions, so there's a mismatch in error handling. Fortunately there usually aren't so many reparse points that the accumulated cost of `STAT()` calls would matter much.

[^1]: That's if `DirEntry` was updated to behave like `os.stat()` and `os.lstat()`, for which `follow_symlinks` is interpreted liberally to include junctions and other name-surrogate file types, not just literal symlinks, while other types of reparse points are always traversed because they're not meant to be handled like links. For example, `DirEntry.is_dir()` incorrectly returns true for a junction that targets a non-existent path.

**Comment by barneygale** [comment]
> A disappointment with these builtin functions is that they can't be leveraged in `pathlib.Path` because they're designed to simply return `True` or `False` without raising `OSError` exceptions. The problem is that the `pathlib.Path` methods `exists()`, `is_dir()`, `is_file()`, and `is_symlink()` only ignore `OSError` exceptions for a small set of `errno` values and Windows error codes.

We can at least use it from `Path.glob()`, which suppresses all `OSError`s. I have a draft PR for that here: https://github.com/python/cpython/pull/117858

I wouldn't mind if we made the `is_` methods suppress all `OSError`s rather than a fairly arbitrary subset as we do now. We don't document what's suppressed, and it's changed before.

File: Lib/test/test_genericpath.py (hunk 5 above)
Comment (nineteendo) [reply Lib/test/test_genericpath.py]: It doesn't, I'll simply add the missing tests..

**Comment by nineteendo** [comment]
Thanks for the feedback! I hope this can be merged now.

**Comment by nineteendo** [comment]
cc @zooba

**Comment by zooba** [comment]
So I'm hesitant to take this for three reasons (and these do apply to previous enhancements as well, but didn't exist at that time):
* `stat` is going to get fast enough on newer Windows for this optimisation to be redundant
* it's a lot of code and not easily maintainable
* I don't know any scenarios where you have a lot of paths and you need to check whether they exist

`isdir` and `islink` are useful to have, because you may `glob` or `listdir` and need to filter its members. But you don't need to `exists` in that case - everything in the directory has to be assumed to exist (and handle the rare race condition when you try to use it).

If someone can show a scenario where you would have a significant (hundreds+) list of paths, need to check whether they exist, but couldn't use one of `isfile`, `isdir` or `islink`, then I could be persuaded on the third point. If there's some reason why this scenario exists for people who can upgrade Python but not update Windows, then I might be convinced on the first point (or alternatively, if lexists doesn't actually get faster with the new stat APIs, which is a possibility).

I don't think we can really reduce the amount of code. If it happened to be shorter and easier to follow then I'd be less concerned about long-term maintenance, but I'm pretty sure it's as good as it gets (without adding indirection and hurting the performance again - same tradeoff we made with the earlier `_path_*` functions).

**Comment by barneygale** [comment]
> If someone can show a scenario where you would have a significant (hundreds+) list of paths, need to check whether they exist, but couldn't use one of isfile, isdir or islink, then I could be persuaded on the third point.

On this specifically, one example would be globbing for `*/__init__.py`. The quickest way to implement that is `os.scandir()` for an initial set of paths, join `__init__.py` onto each of them, and then call `lstat()` to filter out nonexistent paths. This is the approach taken in the pathlib globbing implementation:

https://github.com/python/cpython/blob/a7711a2a4e5cf16b34fc284085da724a8c2c06dd/Lib/glob.py#L508-L520

Note that glob results can include dangling symlinks, hence `lstat()` rather than `stat()`.

**Comment by nineteendo** [comment]
> But I agree that all of these `_path_is*` and `_path_[l]exists` helpers are a lot of code to maintain, taken together. It could be refactored into smaller inline helper functions that can be reused, which would also make the code more readable.

Doesn't seem too difficult for `_path_isdir` & `_path_isfile`:

Screenshot 2024-05-02 at 09 14 05

**Comment by nineteendo** [comment]
Could we simplify this? I already seems to know if it's a directory of file:

https://github.com/python/cpython/blob/a6b610a94bee0e4436aee2825c14f05ec2f22f75/Modules/posixmodule.c#L5123-L5129

```c
slow_path = FALSE;
result = statInfo.FileAttributes & FILE_ATTRIBUTE_DIRECTORY;
```

https://github.com/python/cpython/blob/a6b610a94bee0e4436aee2825c14f05ec2f22f75/Modules/posixmodule.c#L5220-L5226

```c
slow_path = FALSE;
result = !(statInfo.FileAttributes & FILE_ATTRIBUTE_DIRECTORY);
```

**Comment by eryksun** [comment]
> Could we simplify this? It already seems to know if it's a directory of file:

Yes, the fast path can be simplified. If it's not a reparse point, `isdir()` and `isfile()` can always use the entry from the directory listing. That's implemented. `isdir()` can return `False` for a file reparse point. That's implemented. Otherwise it has to ensure that the reparsed target exists. This could be improved. Falling back to the full slow path does more work than necessary since it only needs to check existence via `CreateFileW()`, not the `GetFileInformationByHandleEx()` query. Similarly `isfile()` has to verify existence for a file reparse point.

**Comment by nineteendo** [comment]
Tracking further in #118755.

**Pull request closed without merging**